File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        out = set([])
        nums.sort()
        length = len(nums)
        for i in range(length - 2):

            if i > 0 and nums[i] == nums[i - 1]:
                continue

            target = -nums[i]

            right = length - 1
            left = i + 1

            while right > left:
                if nums[right] + nums[left] > target:
                    right -= 1
                elif nums[right] + nums[left] < target:
                    left += 1
                else: 
                    out.add((nums[i], nums[left], nums[right]))
                    left += 1
                    right -= 1

        return [list(t) for t in out]

File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        int len = nums.size();
        for (int i = 0; i < len - 2; i++) {
            int target = -nums[i];
            if (i > 0 && nums[i] == nums[i - 1]) continue;
            int l = i + 1;
            int r = len - 1;

            while (l < r) {

                if (nums[l] + nums[r] < target) l++;
                else if (nums[l] + nums[r] > target) r--;
                else {
                    res.push_back({nums[i], nums[l], nums[r]});
                    l ++;                 
                    while (l < r && nums[l] == nums[l - 1]) {
                        l ++;
                        continue;
                    }
                }
                    
            }
        }

        return res;
    }
};